### ISSUE TITLE
intel compiler: libsais64 doesn't produce correct SuffixArray

### ISSUE THREAD

**Issue by SGSSGene** [IlyaGrebnov/libsais#20]
Using the Intel compiler and calling the libsais64 does not result in correct suffix array.

## How to reproduce
- Checkout newest commit #d5e74eb93
- <details>
    <summary>create test.c file</summary>

  ```cpp
  #include <stdio.h>
  #include <stdlib.h>
  #include <string.h>
  
  #include <libsais.h>
  #include <libsais64.h>
  
  int main() {
      // input data
      char *Text = "abracadabra";
      int n = strlen(Text);
      int i, j;
  
      // allocate
      int64_t *SA = (int64_t *)malloc(n * sizeof(int64_t));
  
      // sort
      libsais64((unsigned char *)Text, SA, n, 0, NULL);
  
      // output
      for(i = 0; i < n; ++i) {
          printf("SA[%2d] = %2ld: ", i, SA[i]);
          for(j = SA[i]; j < n; ++j) {
              printf("%c", Text[j]);
          }
          printf("$\n");
      }
  
      // deallocate
      free(SA);
  
      return 0;
  }
  ```
</details>

- compiled with `icx -O2 -DNDEBUG src/libsais64.c src/libsais.c test.c`
- run `./a.out`

Output:
```
SA[ 0] = 10: a$                                                                                                                                                
SA[ 1] =  7: abra$                                                                                                                                             
SA[ 2] =  0: abracadabra$                                                                                                                                      
SA[ 3] =  3: acadabra$                                                                                                                                         
SA[ 4] =  0: abracadabra$                                                                                                                                      
SA[ 5] =  0: abracadabra$                                                                                                                                      
SA[ 6] =  3: acadabra$                                                                                                                                         
SA[ 7] =  0: abracadabra$                                                                                                                                      
SA[ 8] =  0: abracadabra$                                                                                                                                      
SA[ 9] =  0: abracadabra$                                                                                                                                      
SA[10] =  0: abracadabra$       
```

Expected Output:
```
SA[ 0] = 10: a$
SA[ 1] =  7: abra$
SA[ 2] =  0: abracadabra$
SA[ 3] =  3: acadabra$
SA[ 4] =  5: adabra$
SA[ 5] =  8: bra$
SA[ 6] =  1: bracadabra$
SA[ 7] =  4: cadabra$
SA[ 8] =  6: dabra$
SA[ 9] =  9: ra$
SA[10] =  2: racadabra$
```

# Version
`icx --version`
```
Intel(R) oneAPI DPC++/C++ Compiler 2023.2.0 (2023.2.0.20230622)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/intel/oneapi/compiler/2023.2.0/linux/bin-llvm
Configuration file: /opt/intel/oneapi/compiler/2023.2.0/linux/bin-llvm/../bin/icx.cfg
```

Things I know:
- using `-O1` works, code works with -O2 under clang and gcc.
- `libsais` works but `libsais64` doesn't

Thank you for all the help!

**Comment by eseiler** [comment]
**TLDR**: Compiles with `-no-ansi-alias` (aka `-fno-strict-aliasing`)

What follows is just what I found out while debugging:

For this example, the code path ends up in

https://github.com/IlyaGrebnov/libsais/blob/d5e74eb93e6d8dcf3b753a2818346206be1f5fc5/src/libsais64.c#L6232-L6235

The compiler may assume that `uint32_t * S` and `uint64_t * D` do not alias and hence are different objects, and IntelLLVM takes advantage of that.

IntelLLVM will vectorize the for-loop:

```
--- !Passed
Pass:            slp-vectorizer
Name:            StoresVectorized
DebugLoc:        { File: 'src/libsais64.c', Line: 6234, Column: 114 }
Function:        libsais64
Args:
  - String:          'Stores SLP vectorized with cost '
  - Cost:            '-1'
  - String:          ' and with tree size '
  - TreeSize:        '3'
...
```

Hence, the output is `[10, 7, 0, ...]`.

When disabling the vectorization via `-fno-slp-vectorize`, the output is `[10, 0, 0, ..., 0]`.

The first value in `SA`, which is accessed via `S/D`, is `30064771082`, which is `011100000000000000000000000000001010` in binary.
The first 32 bit are cast to `uint64_t` and written. `00.001010` is the `10` we see in the output. The rest is overwritten with 0s.
I'm assuming that the compiler also reorders the loop to go from left to right, I haven't looked at the assembly.

When explicitly marking the pointers as aliasing, e.g.,
```cpp
typedef __attribute__ ((may_alias)) uint32_t uint32_alias_t;
typedef __attribute__ ((may_alias)) uint64_t uint64_alias_t;

static void libsais64_convert_right_to_left_32u_to_64u(uint32_alias_t * S, uint64_alias_t * D, fast_sint_t omp_block_start, fast_sint_t omp_block_size)
{
    fast_sint_t i, j; for (i = omp_block_start + omp_block_size - 1, j = omp_block_start; i >= j; i -= 1) { D[i] = (uint64_t)S[i]; }
}
```
the code works as expected:

```
--- !Missed
Pass:            gvn
Name:            LoadClobbered
DebugLoc:        { File: 'src/libsais64.c', Line: 6240, Column: 126 }
Function:        libsais64
Args:
  - String:          'load of type '
  - Type:            ''
  - String:          ' not eliminated'
  - String:          ' because it is clobbered by '
  - ClobberedBy:     store
    DebugLoc:        { File: 'src/libsais64.c', Line: 6234, Column: 114 }
...
```

the same can be achieved with
```cpp
static void libsais64_convert_right_to_left_32u_to_64u(uint32_alias_t * S, uint64_alias_t * D, fast_sint_t omp_block_start, fast_sint_t omp_block_size)
{
    volatile fast_sint_t i;
    fast_sint_t j; for (i = omp_block_start + omp_block_size - 1, j = omp_block_start; i >= j; i -= 1) { D[i] = (uint64_t)S[i]; }
}
```
then the loop has to be evaluated as written because no assumptions about `i` in `[i]` can be made.

There are other places where strict-aliasing is violated, which are not hit by the example.
Even though Clang/GCC enable strict-aliasing with (at least) O2 and above, it seems like IntelLLVM's more aggressive optimizations make the difference in this case.

**Comment by IlyaGrebnov** [comment]
I agree that, from a strict-aliasing rules perspective, my code is not compliant and I will make the necessary changes to correct this. I am looking for a solution that is more compiler-independent.

**Comment by IlyaGrebnov** [comment]
Ok. This should be fixed in latest version. I also confirmed that @eseiler is correct on assumption. The issue is that compiler reordered the loop to go from left to right. Unfortunately with the fix I was not able to preserve vectorization as I have to opt it to more compiler-independent approach.
